services/shhext/mailservers/connmanager.go

```diff
@@ -106,19 +108,24 @@ func (ps *ConnectionManager) Start() {
 			case ev := <-events:
 				processPeerEvent(state, ev)
 			case ev := <-whisperEvents:
-				// TODO what about completed but with error? what about expired envelopes?
+				// TODO treat failued request the same way as expired
 				switch ev.Event {
 				case whisper.EventMailServerRequestSent:
 					requests[ev.Hash] = struct{}{}
 				case whisper.EventMailServerRequestCompleted:
+					// reset failures count on first success
+					failuresPerServer[ev.Peer] = 0
 					delete(requests, ev.Hash)
 				case whisper.EventMailServerRequestExpired:
 					_, exist := requests[ev.Hash]
 					if !exist {
 						continue
 					}
+					failuresPerServer[ev.Peer]++
 					log.Debug("request to a mail server expired, disconncet a peer", "address", ev.Peer)
-					state.nodeDisconnected(ev.Peer)
+					if failuresPerServer[ev.Peer] == ps.maxFailures {
```

Comment:
I would always use `>=` here for future-proofing (if we add `failuresPerServer[ev.Peer]++` somewhere else or by mistake)

Reply:
Or even better set an invariant which will check and notify (`log.Error`?) in case `failuresPerServer[ev.Peer] > ps.maxFailures`.

Reply:
yeap, that might be a good idea